initFabian/Cely
# Add biometrics

## Purpose:
Add biometrics functionality. 

## New Cely Credentials API

Please refer to [documentation docs](http://celylog.in/api/cely-credentials/) for usage. I know I pulled the trigger a little early around the documentation site, **buuuttttt** I wanted to focus on the experience around documentation and have that drive the code for this release. Please give your honest opinion and API design and potential shortcomings. 

## Cely Documentation

This is going to be something new and entirely different so don't feel responsible for this part of the code. It uses [`mkdocs`](https://www.mkdocs.org/) to generate the site. But if you are able to check out the docs and give your thoughts that would be amazing! I understand all this was made in a cave and should be viewed as a First Draft.

## What did I do?

- Introduces new `CelyCredentials` API
- Added new file demo file `CelyDemo/BiometricsLoginViewController.swift` that in the **Cely Demo Target** that uses new `CelyCredentials` API
- set `CelyDemo/Appdelegate.swift` to point to `BiometricsLoginViewController.swift` for testing
- Add a button to `CelyDemo/SettingsViewController.swift` that logs credentials to the console
- improve encapsulation (private) around methods/variables throughout Cely
- Replace `StorageResult` with Swift `Result` 
- replace `CelyStorageProtocol.removeAllData()` with `clearStorage()`
- Add class `CelyKeychain`
  - responsible for CRUD operations for `KeychainObject` (Keychain Items)
- Add class `CelySecureStorage`
  - responsible for creating `KeychainObject` and storing them with `CelyKeychain`
  - implements `CelyStorageProtocol` (`.clearStorage()` gets called from `CelyStorage` when user is logging out)
- Add class `KeychainObject`
  - syntactic sugar around `KeychainItem`. (I'm not entirely in love with how this came out around `toGetMap`/`toLookupMap` since they aren't really descriptive on what they're used for.)

## Notes


When I created this PR, it was to finally get the code in front of someone else. In terms of what this PR sought out to do (add the biometrics functionality), it does just that. but as far as how we get there, that's where I have thoughts. Things such as should we allow the developer to use `kSecClassGenericPassword` instead of `kSecClassInternetPassword` or use `kSecAttrAccessibleWhenUnlocked` instead of `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`. Definitely read Cely's documentation over [Understanding Keychain](http://celylog.in/guides/understanding_keychain/) to form an opinion on what Cely should/shouldn't do.

In all, I would say this PR is 97% done. Really the only thing I don't like is around `KeychainObject`, but that class is only used internally and not exposed to the user/developer. so I'm not sure if we want to give that class the green light for now and just improve it at a later time.


Lastly, this PR did a lot. I really wish it didn't, but I wanted to do one big push for v3 to get the ball rolling. 

## How to Test:

0. Have device with biometrics (FaceID/TouchID)
1. Pull down branch
2. Change Scheme to Cely Demo
3. Error will happen around team/bundleID. change values to allow App to run
  - if you can help with find a long term solution on this, that would be awesome!
4. Log into app with the credentials **username:asdf** and **password:asdf**
5. **VERIFY**: App logs you in
6. go to settings page and click _**Get Credentials**_
7. **VERIFY**: biometrics are run on device
8. **VERIFY**: credentials are logged to the console.
